core/src/mast/mod.rs

```diff
@@ -370,48 +357,73 @@ impl MastForest {
     }
 
     pub fn decorators(&self) -> &[Decorator] {
-        self.decorators.as_slice()
+        self.debug_info.decorators()
     }
 
     /// Returns decorator indices for a specific operation within a node.
     ///
     /// This is the primary accessor for reading decorators from the centralized storage.
     /// Returns a slice of decorator IDs for the given operation.
-    #[inline(always)]
+    #[inline]
     pub fn decorator_indices_for_op(
         &self,
         node_id: MastNodeId,
         local_op_idx: usize,
     ) -> &[DecoratorId] {
-        self.op_decorator_storage
-            .decorator_ids_for_operation(node_id, local_op_idx)
-            .unwrap_or(&[])
+        self.debug_info.decorators_for_operation(node_id, local_op_idx)
     }
 
     /// Returns an iterator over decorator references for a specific operation within a node.
     ///
     /// This is the preferred method for accessing decorators, as it provides direct
     /// references to the decorator objects.
-    #[inline(always)]
+    #[inline]
     pub fn decorators_for_op<'a>(
         &'a self,
         node_id: MastNodeId,
         local_op_idx: usize,
     ) -> impl Iterator<Item = &'a Decorator> + 'a {
         self.decorator_indices_for_op(node_id, local_op_idx)
             .iter()
-            .map(move |&decorator_id| &self.decorators[decorator_id])
+            .map(move |&decorator_id| &self[decorator_id])
     }
 
     /// Returns decorator links for a node, including operation indices.
     ///
     /// This provides a flattened view of all decorators for a node with their operation indices.
-    #[inline(always)]
+    #[inline]
     pub fn decorator_links_for_node<'a>(
         &'a self,
         node_id: MastNodeId,
     ) -> Result<DecoratedLinks<'a>, DecoratorIndexError> {
-        self.op_decorator_storage.decorator_links_for_node(node_id)
+        self.debug_info.decorator_links_for_node(node_id)
+    }
+
+    /// Returns the decorators to be executed before this node is executed.
+    #[inline]
+    pub fn before_enter_decorators(&self, node_id: MastNodeId) -> &[DecoratorId] {
+        self.debug_info.before_enter_decorators(node_id)
+    }
+
+    /// Returns the decorators to be executed after this node is executed.
+    #[inline]
+    pub fn after_exit_decorators(&self, node_id: MastNodeId) -> &[DecoratorId] {
+        self.debug_info.after_exit_decorators(node_id)
+    }
+
+    /// Adds decorator IDs for a node to the storage.
+    /// Used when building nodes for efficient decorator access during execution.
+    pub(crate) fn add_node_decorators(
+        &mut self,
+        node_id: MastNodeId,
+        before_enter: &[DecoratorId],
+        after_exit: &[DecoratorId],
+    ) {
+        self.debug_info.node_decorator_storage_mut().add_node_decorators(
+            node_id,
+            before_enter,
+            after_exit,
+        );
     }
```

Comment:
We use this function only in `SplitNode::add_to_forest()` in other places (e.g., `JoinNode::add_to_forest()` we access `debug_info` directly. I think we should make these things consistent:

- We either get rid of this function, or:
- Start using this in all places instead of accessing `debug_info` directly.

Also, the implementation of this function could be just:
```rust
self.debug_info.add_node_decorators(node_id, before_enter, after_exit);
```

Reply:
Used everywhere